mirrord-layer/tests/integration.rs

```diff
@@ -0,0 +1,149 @@
+use std::{collections::HashMap, process::Stdio, time::Duration};
+
+use actix_codec::Framed;
+use futures::{stream::StreamExt, SinkExt};
+use mirrord_protocol::{
+    tcp::{DaemonTcp, LayerTcp, NewTcpConnection, TcpData},
+    ClientMessage, DaemonCodec, DaemonMessage,
+};
+use rstest::rstest;
+use tokio::{
+    io::AsyncWriteExt,
+    net::{TcpListener, TcpStream},
+    process::Command,
+};
+
+/// Accept a connection from the libraries and verifies the first message it is supposed to send
+/// to the agent - GetEnvVarsRequest. Send back a response.
+/// Return the codec of the accepted stream.
+async fn accept_library_connection(listener: &TcpListener) -> Framed<TcpStream, DaemonCodec> {
+    let (stream, _) = listener.accept().await.unwrap();
+    let mut codec = Framed::new(stream, DaemonCodec::new());
+    let msg = codec.next().await.unwrap().unwrap();
+    if let ClientMessage::GetEnvVarsRequest(request) = msg {
+        assert!(request.env_vars_filter.is_empty());
+        assert_eq!(request.env_vars_select.len(), 1);
+        assert!(request.env_vars_select.contains("*"));
+    } else {
+        panic!("unexpected request {:?}", msg)
+    }
+    codec
+        .send(DaemonMessage::GetEnvVarsResponse(Ok(HashMap::new())))
+        .await
+        .unwrap();
+    codec
+}
+
+/// Send the layer a message telling it the target got a new incoming connection.
+/// There is no such actual connection, because there is no target, but the layer should start
+/// a mirror connection with the application.
+async fn send_new_connection(codec: &mut Framed<TcpStream, DaemonCodec>, connection_id: u64) {
+    codec
+        .send(DaemonMessage::Tcp(DaemonTcp::NewConnection(
+            NewTcpConnection {
+                connection_id,
+                address: "127.0.0.1".parse().unwrap(),
+                destination_port: "80".parse().unwrap(),
+                source_port: "31415".parse().unwrap(),
+            },
+        )))
+        .await
+        .unwrap();
+}
+
+async fn send_tcp_data(
+    codec: &mut Framed<TcpStream, DaemonCodec>,
+    message_data: &str,
+    connection_id: u64,
+) {
+    codec
+        .send(DaemonMessage::Tcp(DaemonTcp::Data(TcpData {
+            connection_id,
+            bytes: Vec::from(message_data),
+        })))
+        .await
+        .unwrap();
+}
+
+/// Tell the layer there is a new incoming connection, then send data "from that connection".
+async fn send_connection_then_data(
+    codec: &mut Framed<TcpStream, DaemonCodec>,
+    message_data: &str,
+    connection_id: u64,
+) {
+    send_new_connection(codec, connection_id).await;
+    send_tcp_data(codec, message_data, connection_id).await;
+}
+
```

Comment:
Idea for improvement - create a struct that can be created same as `accept_library_connection` then have methods such as above (based on the shared requirement for having codec)

Reply:
@aviramha like this?

Reply:
Yup. Please note that in Rust you don't have to type self
```rs
async fn send_new_connection(self: &mut Self) -> u64
```
can be written as
```rs
async fn send_new_connection(&mut self) -> u64
```